packages/filesystem/src/browser-only/opfs-filesystem-provider.ts

```diff
@@ -0,0 +1,345 @@
+                        return Promise.resolve(true);
+                    }
+                    return Reflect.get(target, prop, receiver);
+                }
+            }));
+            return true;
+        };
+        this.initialized = setup();
+    }
+
+    watch(_resource: URI, _opts: WatchOptions): Disposable {
+        return Disposable.NULL;
+    }
+    // MARK: stat
+    async stat(resource: URI): Promise<Stat> {
+        try {
+            await this.initialized;
+
+            const handle = await this.toFileSystemHandle(resource);
+
+            if (handle.kind === 'file') {
+                const fileHandle = handle as FileSystemFileHandle;
+                const file = await fileHandle.getFile();
+                return {
+                    type: FileType.File,
+                    ctime: file.lastModified,
+                    mtime: file.lastModified,
+                    size: file.size
+                };
+            } else if (handle.kind === 'directory') {
+                return {
+                    type: FileType.Directory,
+                    ctime: 0,
+                    mtime: 0,
+                    size: 0
+                };
+            }
+
+            throw createFileSystemProviderError('File does not exist', FileSystemProviderErrorCode.FileNotFound);
+
+        } catch (error) {
+            throw createFileSystemProviderError('File does not exist', FileSystemProviderErrorCode.FileNotFound);
+        }
+    }
+    // MARK: mkdir
+    async mkdir(resource: URI): Promise<void> {
+        await this.initialized;
+        try {
+            await this.toFileSystemHandle(resource, { create: true, isDirectory: true });
+            this.onDidChangeFileEmitter.fire([{ resource, type: FileChangeType.ADDED }]);
+        } catch (error) {
+            throw toFileSystemProviderError(error, true);
+        }
+    }
+
+    // MARK: readdir
+    async readdir(resource: URI): Promise<[string, FileType][]> {
+        await this.initialized;
+
+        try {
+            // Get the directory handle from the directoryHandle
+            const directoryHandle = await this.toFileSystemHandle(resource, { create: false, isDirectory: true }) as FileSystemDirectoryHandle;
+
+            const result: [string, FileType][] = [];
+
+            // Iterate through the entries in the directory (files and subdirectories)
+            for await (const [name, handle] of directoryHandle.entries()) {
+                try {
+                    // Determine the type of the entry (file or directory)
+                    if (handle.kind === 'file') {
+                        result.push([name, FileType.File]);
+                    } else if (handle.kind === 'directory') {
+                        result.push([name, FileType.Directory]);
+                    }
+                } catch (error) {
+                    // Ignore errors for individual entries, log them and continue
+                    // console.error(error);
```

Comment:
Are any actual errors thrown here? What does this try-catch protect against?

Reply:
You're right. The type system should protect us against any other type than file and directory